pkg/agent/ipassigner/responder/ndp_responder.go

```diff
@@ -0,0 +1,230 @@
+// limitations under the License.
+
+package responder
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/mdlayher/ndp"
+	"golang.org/x/net/ipv6"
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+)
+
+const (
+	solicitedNodeMulticastAddressPrefix = "ff02::1:ff00:0"
+)
+
+type ndpConn interface {
+	WriteTo(message ndp.Message, cm *ipv6.ControlMessage, dstIP net.IP) error
+	ReadFrom() (ndp.Message, *ipv6.ControlMessage, net.IP, error)
+	JoinGroup(net.IP) error
+	LeaveGroup(net.IP) error
+	Close() error
+}
+
+func parseIPv6SolicitedNodeMulticastAddress(ip net.IP) (net.IP, int) {
+	group := net.ParseIP(solicitedNodeMulticastAddressPrefix)
+	// copy lower 24 bits
+	copy(group[13:], ip[13:])
+	key := int(group[13])<<16 | int(group[14])<<8 | int(group[15])
+	return group, key
+}
+
+type ndpResponder struct {
+	iface           *net.Interface
+	conn            ndpConn
+	closed          chan struct{}
+	assignedIPs     map[string]struct{}
+	multicastGroups map[int]int
+	mutex           sync.Mutex
+}
+
+var _ Responder = (*ndpResponder)(nil)
+
+func NewNDPResponder(iface *net.Interface) (*ndpResponder, error) {
+	conn, _, err := ndp.Listen(iface, ndp.LinkLocal)
+	if err != nil {
+		return nil, err
+	}
+	return &ndpResponder{
+		iface:           iface,
+		conn:            conn,
+		closed:          make(chan struct{}),
+		multicastGroups: make(map[int]int),
+		assignedIPs:     map[string]struct{}{},
+	}, nil
+}
+
+func (r *ndpResponder) InterfaceName() string {
+	return r.iface.Name
+}
+
+func (r *ndpResponder) Advertise(ip net.IP) error {
+	select {
+	case <-r.closed:
+		return fmt.Errorf("NDP responder socket is closed")
+	default:
+		if !utilnet.IsIPv6(ip) {
+			return fmt.Errorf("only IPv6 is supported")
+		}
+		na := &ndp.NeighborAdvertisement{
+			Override:      true,
+			TargetAddress: ip,
+			Options: []ndp.Option{
+				&ndp.LinkLayerAddress{
+					Direction: ndp.Target,
+					Addr:      r.iface.HardwareAddr,
+				},
+			},
+		}
+		return r.conn.WriteTo(na, nil, net.IPv6linklocalallnodes)
+	}
+}
+
+func (r *ndpResponder) handleNeighborSolicitation() error {
+	pkt, _, srcIP, err := r.conn.ReadFrom()
+	if err != nil {
+		return err
+	}
+	ns, ok := pkt.(*ndp.NeighborSolicitation)
+	if !ok {
+		return nil
+	}
+	var nsSourceHWAddr net.HardwareAddr
+	for _, o := range ns.Options {
+		addr, ok := o.(*ndp.LinkLayerAddress)
+		if !ok {
+			continue
+		}
+		if addr.Direction != ndp.Source {
+			continue
+		}
+		nsSourceHWAddr = addr.Addr
+		break
+	}
+	if nsSourceHWAddr == nil {
+		return nil
+	}
+
+	if _, ok := r.assignedIPs[ns.TargetAddress.String()]; !ok {
+		klog.V(4).InfoS("Ignore Neighbor Solicitation", "ip", ns.TargetAddress.String(), "interface", r.iface.Name)
+		return nil
+	}
+	na := &ndp.NeighborAdvertisement{
+		Solicited:     true,
+		TargetAddress: ns.TargetAddress,
+		Options: []ndp.Option{
+			&ndp.LinkLayerAddress{
+				Direction: ndp.Target,
+				Addr:      r.iface.HardwareAddr,
+			},
+		},
+	}
```

Comment:
Doesn't it need to use `nsSourceHWAddr` to set destination address?

Reply:
The message is composed according to https://datatracker.ietf.org/doc/html/rfc4861#section-4.4. We only need the src IP as the destination address to send the packet in `r.conn.WriteTo(na, nil, srcIP)`.